src/compiler/checker.ts

```diff
@@ -11134,7 +11121,7 @@ namespace ts {
                 case SyntaxKind.TypeReference:
                     return getTypeFromTypeReference(<TypeReferenceNode>node);
                 case SyntaxKind.TypePredicate:
-                    return booleanType;
+                    return (<TypePredicateNode>node).assertsModifier ? voidType : booleanType;

```

Comment:
I, too, would also still love to see a non-void returning `asserts` function. Having the shorthand for the `void`-returning common case is fine, but I do think there's plenty of uses (see: every other petitioner in this PR) for a "returning" assertion (especially if in the future we maybe bind them in expression positions). Syntax for that is 🚲 🏠 able all day long, though I definitely think there's some overlap with the intent of `typeof` when used within a conditional.

Reply:
The case for non-void returning `asserts` functions is much less compelling and not part of the motivating scenarios. I don't want to go there in this PR.